File: partner-kit/node/selftest.js
"use strict";

/**
 * Tiket Connect self-test — verifies your deployment end-to-end WITHOUT
 * involving TIKET. Boots the adapter on a local port with demo data, then
 * fires signed and unsigned requests at it and checks every answer.
 *
 *   node selftest.js
 *
 * All checks green → the kit itself is working; deploy it wired to your real
 * database and send TIKET the public URL + your shared secret.
 */

const assert = require("assert");
const {
  createTiketConnect,
  computeSignature,
  compareClaimToRecord,
  normalizers,
  KIT_VERSION,
  API_VERSION,
} = require("./tiket-connect");

// Shared protocol spec vectors — the Go agent's selftest consumes the SAME
// file. If this suite disagrees with the vectors, fix the implementation or
// change the vectors first (never silently diverge).
const SPEC_VECTORS = require("../spec/test-vectors.json");

const SECRET = "selftest-secret-0123456789abcdef";

// Demo "database": one seated ticket, one standing, one cancelled.
const DEMO_TICKETS = {
  "1000000000001": {
    barcode: "1000000000001",
    event_name: "עומר אדם — סיבוב קיץ",
    venue: "היכל מנורה מבטחים",
    date: "2030-08-15",
    time: "21:00",
    section: "A",
    row: "12",
    seat: "7",
    status: "active",
    original_price: 350,
    ticket_ref: "DEMO-1",
  },
  "1000000000002": {
    barcode: "1000000000002",
    event_name: "נועה קירל",
    venue: "פארק הירקון",
    date: "2030-09-01",
    time: "20:30",
    is_standing: true,
    status: "active",
    original_price: 280,
    ticket_ref: "DEMO-2",
  },
  "1000000000003": {
    barcode: "1000000000003",
    event_name: "הופעה שבוטלה",
    venue: "היכל התרבות",
    date: "2030-10-10",
    time: "20:00",
    is_standing: true,
    status: "cancelled",
    ticket_ref: "DEMO-3",
  },
};

async function post(baseUrl, path, payload, { sign = true, timestamp } = {}) {
  const rawBody = JSON.stringify(payload);
  const ts = timestamp ?? Math.floor(Date.now() / 1000);
  const headers = { "Content-Type": "application/json" };
  if (sign) {
    headers["X-Tiket-Timestamp"] = String(ts);
    headers["X-Tiket-Signature"] = computeSignature(SECRET, ts, rawBody);
  }
  const res = await fetch(`${baseUrl}${path}`, { method: "POST", headers, body: rawBody });
  return { status: res.status, body: await res.json() };
}

function claim(overrides) {
  return {
    api_version: API_VERSION,
    request_id: `selftest-${Math.random().toString(36).slice(2)}`,
    ticket: {
      barcode: "1000000000001",
      event_name: "עומר אדם — סיבוב קיץ",
      artist: "עומר אדם",
      venue: "היכל מנורה",
      date: "2030-08-15",
      time: "21:00",
      section: "A",
      row: "12",
      seat: "7",
      is_standing: false,
      ...overrides,
    },
  };
}

async function main() {
  let transferCalls = 0;
  const connect = createTiketConnect({
    secret: SECRET,
    logger: null,
    lookupTicket: (barcode) => DEMO_TICKETS[barcode] || null,
    transferTicket: (transfer) => {
      transferCalls++;
      return {
        ok: true,
        new_barcode: `REISSUED-${transfer.transfer_ref}`,
        barcode_format: "qr",
        new_ticket_ref: "DEMO-1-B",
        delivery: "new_barcode",
      };
    },
  });

  const server = connect.listen(0);
  await new Promise((resolve) => server.on("listening", resolve));
  const baseUrl = `http://127.0.0.1:${server.address().port}`;

  let passed = 0;
  const check = (name, fn) => {
    try {
      fn();
      passed++;
      console.log(`  ✓ ${name}`);
    } catch (err) {
      console.error(`  ✗ ${name}\n    ${err.message}`);
      process.exitCode = 1;
    }
  };

  console.log(`Tiket Connect self-test (kit v${KIT_VERSION}, API v${API_VERSION})\n`);

  // 1. Health
  const health = await fetch(`${baseUrl}/tiket/health`).then((r) => r.json());
  check("health endpoint answers", () => assert.strictEqual(health.ok, true));

  // 2. Perfect match
  let r = await post(baseUrl, "/tiket/verify", claim());
  check("valid ticket → match", () => {
    assert.strictEqual(r.status, 200);
    assert.strictEqual(r.body.result, "match");
    assert.ok(r.body.confidence >= 90, `confidence ${r.body.confidence} < 90`);
    assert.strictEqual(r.body.original_price, 350);
    assert.strictEqual(r.body.ticket_ref, "DEMO-1");
  });

  // 3. Fuzzy tolerance: partial venue name + padded seat numbers still match
  r = await post(baseUrl, "/tiket/verify", claim({ venue: "מנורה מבטחים", row: "012", seat: "07" }));
  check("fuzzy venue/seat formats → match", () => {
    assert.strictEqual(r.body.result, "match");
  });

  // 4. Wrong date → mismatch, date listed as unmatched
  r = await post(baseUrl, "/tiket/verify", claim({ date: "2030-08-16" }));
  check("wrong date → mismatch", () => {
    assert.strictEqual(r.body.result, "mismatch");
    assert.ok(r.body.unmatched_fields.includes("date"));
  });

  // 5. Unknown barcode → not_found
  r = await post(baseUrl, "/tiket/verify", claim({ barcode: "9999999999999" }));
  check("unknown barcode → not_found", () => {
    assert.strictEqual(r.body.result, "not_found");
  });

  // 6. Standing ticket matches without seat details
  r = await post(baseUrl, "/tiket/verify", {
    ...claim({
      barcode: "1000000000002",
      event_name: "נועה קירל",
      artist: "נועה קירל",
      venue: "פארק הירקון",
      date: "2030-09-01",
      time: "20:30",
      section: "",
      row: "",
      seat: "",
      is_standing: true,
    }),
  });
  check("standing ticket → match", () => assert.strictEqual(r.body.result, "match"));

  // 7. Cancelled ticket reports its status
  r = await post(baseUrl, "/tiket/verify", claim({
    barcode: "1000000000003",
    event_name: "הופעה שבוטלה",
    artist: "הופעה שבוטלה",
    venue: "היכל התרבות",
    date: "2030-10-10",
    time: "20:00",
    section: "", row: "", seat: "",
    is_standing: true,
  }));
  check("cancelled ticket → ticket_status=cancelled", () => {
    assert.strictEqual(r.body.ticket_status, "cancelled");
  });

  // 8. Missing signature → 401
  r = await post(baseUrl, "/tiket/verify", claim(), { sign: false });
  check("unsigned request → 401", () => assert.strictEqual(r.status, 401));

  // 9. Stale timestamp → 401 (replay protection)
  r = await post(baseUrl, "/tiket/verify", claim(), { timestamp: Math.floor(Date.now() / 1000) - 3600 });
  check("stale timestamp → 401", () => assert.strictEqual(r.status, 401));

  // 10. Tampered body → 401
  {
    const payload = claim();
    const rawBody = JSON.stringify(payload);
    const ts = Math.floor(Date.now() / 1000);
    const sig = computeSignature(SECRET, ts, rawBody);
    const tampered = rawBody.replace("1000000000001", "1000000000002");
    const res = await fetch(`${baseUrl}/tiket/verify`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Tiket-Timestamp": String(ts),
        "X-Tiket-Signature": sig,
      },
      body: tampered,
    });
    check("tampered body → 401", () => assert.strictEqual(res.status, 401));
  }

  // ── Ownership transfer ──

  const transferPayload = (overrides) => ({
    api_version: API_VERSION,
    request_id: `selftest-${Math.random().toString(36).slice(2)}`,
    transfer: {
      barcode: "1000000000001",
      ticket_ref: "DEMO-1",
      transfer_ref: "TIKET-TX-0001",
      new_holder: { first_name: "דנה", last_name: "כהן", email: "buyer@example.com", phone: "0501234567" },
      ...overrides,
    },
  });

  // 11. Health advertises transfer support
  const health2 = await fetch(`${baseUrl}/tiket/health`).then((x) => x.json());
  check("health advertises transfer_supported", () =>
    assert.strictEqual(health2.transfer_supported, true)
  );

  // 12. Valid transfer → transferred + reissued barcode (with its symbology)
  r = await post(baseUrl, "/tiket/transfer", transferPayload());
  check("valid transfer → transferred", () => {
    assert.strictEqual(r.status, 200);
    assert.strictEqual(r.body.result, "transferred");
    assert.strictEqual(r.body.new_barcode, "REISSUED-TIKET-TX-0001");
    assert.strictEqual(r.body.barcode_format, "qr");
    assert.strictEqual(r.body.old_barcode_invalidated, true);
  });

  // 13. Replayed transfer_ref → same answer, no second transfer
  r = await post(baseUrl, "/tiket/transfer", transferPayload());
  check("replayed transfer_ref → idempotent", () => {
    assert.strictEqual(r.body.result, "transferred");
    assert.strictEqual(r.body.new_barcode, "REISSUED-TIKET-TX-0001");
    assert.strictEqual(transferCalls, 1, `transferTicket ran ${transferCalls} times, expected 1`);
  });

  // 14. Transfer of unknown barcode → not_found
  r = await post(baseUrl, "/tiket/transfer", transferPayload({ barcode: "9999999999999", transfer_ref: "TIKET-TX-0002" }));
  check("transfer unknown barcode → not_found", () => assert.strictEqual(r.body.result, "not_found"));

  // 15. Transfer of a cancelled ticket → rejected before your code runs
  r = await post(baseUrl, "/tiket/transfer", transferPayload({ barcode: "1000000000003", transfer_ref: "TIKET-TX-0003" }));
  check("transfer cancelled ticket → rejected", () => {
    assert.strictEqual(r.body.result, "rejected");
    assert.ok(String(r.body.reason).startsWith("ticket_not_active"));
  });

  // 16. Unsigned transfer → 401
  r = await post(baseUrl, "/tiket/transfer", transferPayload({ transfer_ref: "TIKET-TX-0004" }), { sign: false });
  check("unsigned transfer → 401", () => assert.strictEqual(r.status, 401));

  // 17. Adapter without transferTicket answers not_supported
  {
    const bare = createTiketConnect({
      secret: SECRET,
      logger: null,
      lookupTicket: (barcode) => DEMO_TICKETS[barcode] || null,
    });
    const bareServer = bare.listen(0);
    await new Promise((resolve) => bareServer.on("listening", resolve));
    const bareUrl = `http://127.0.0.1:${bareServer.address().port}`;
    const bareHealth = await fetch(`${bareUrl}/tiket/health`).then((x) => x.json());
    r = await post(bareUrl, "/tiket/transfer", transferPayload({ transfer_ref: "TIKET-TX-0005" }));
    check("no transferTicket → not_supported (+health says so)", () => {
      assert.strictEqual(bareHealth.transfer_supported, false);
      assert.strictEqual(r.body.result, "not_supported");
    });
    bareServer.close();
  }

  // ── Key rotation (X-Tiket-Key-Id) ──

  {
    const SECRET_B = "selftest-secret-B-fedcba9876543210";
    const rotating = createTiketConnect({
      secrets: { k_old: SECRET, k_new: SECRET_B },
      logger: null,
      lookupTicket: (barcode) => DEMO_TICKETS[barcode] || null,
    });
    const rotServer = rotating.listen(0);
    await new Promise((resolve) => rotServer.on("listening", resolve));
    const rotUrl = `http://127.0.0.1:${rotServer.address().port}`;

    const signedWith = async (secret, keyId) => {
      const rawBody = JSON.stringify(claim());
      const ts = Math.floor(Date.now() / 1000);
      const headers = {
        "Content-Type": "application/json",
        "X-Tiket-Timestamp": String(ts),
        "X-Tiket-Signature": computeSignature(secret, ts, rawBody),
      };
      if (keyId) headers["X-Tiket-Key-Id"] = keyId;
      const res = await fetch(`${rotUrl}/tiket/verify`, { method: "POST", headers, body: rawBody });
      return res.status;
    };

    // 18. Key id selects the right secret
    const s1 = await signedWith(SECRET_B, "k_new");
    const s2 = await signedWith(SECRET, "k_old");
    check("18: signed with either configured key (+key-id) → 200", () => {
      assert.strictEqual(s1, 200);
      assert.strictEqual(s2, 200);
    });

    // 19. Missing key-id header falls back to trying all configured secrets
    const s3 = await signedWith(SECRET_B, null);
    check("19: no key-id header → all secrets tried → 200", () => assert.strictEqual(s3, 200));

    // 20. Wrong secret still rejected, with or without a key-id
    const s4 = await signedWith("wrong-secret-000000000000000000", "k_new");
    check("20: wrong secret → 401", () => assert.strictEqual(s4, 401));

    rotServer.close();
  }

  // ── Shared spec vectors (partner-kit/spec/test-vectors.json) ──

  check("21: spec signature vectors", () => {
    for (const v of SPEC_VECTORS.signature) {
      assert.strictEqual(
        computeSignature(v.secret, v.timestamp, v.body),
        v.expected_signature,
        v.name || v.key_id
      );
    }
  });

  check("22: spec normalization vectors", () => {
    const impls = {
      date: normalizers.normDate,
      time: normalizers.normTime,
      text: normalizers.normText,
      seat: normalizers.normSeatPart,
    };
    for (const v of SPEC_VECTORS.normalization) {
      assert.strictEqual(impls[v.fn](v.input), v.expected, `${v.fn}(${JSON.stringify(v.input)})`);
    }
  });

  check("23: spec comparison vectors", () => {
    for (const v of SPEC_VECTORS.comparison) {
      const r = compareClaimToRecord(v.claim, v.record);
      const result =
        r.confidence >= SPEC_VECTORS.match_threshold && !r.criticalFailed ? "match" : "mismatch";
      assert.strictEqual(r.confidence, v.expected.confidence, `${v.name}: confidence`);
      assert.deepStrictEqual(r.matched, v.expected.matched_fields, `${v.name}: matched`);
      assert.deepStrictEqual(r.unmatched, v.expected.unmatched_fields, `${v.name}: unmatched`);
      assert.strictEqual(r.criticalFailed, v.expected.critical_failed, `${v.name}: critical`);
      assert.strictEqual(result, v.expected.result, `${v.name}: result`);
    }
  });

  const TOTAL = 23;
  server.close();
  console.log(`\n${passed}/${TOTAL} checks passed${passed === TOTAL ? " — the kit is ready 🎉" : ""}`);
}

main().catch((err) => {
  console.error("self-test crashed:", err);
  process.exit(1);
});

File: partner-kit/agent/internal/selftest/selftest.go
// Package selftest validates an agent build with no TIKET involvement:
// the shared protocol vectors (partner-kit/spec/test-vectors.json) plus a
// live in-process loop over the verify/transfer handlers.
package selftest

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"reflect"
	"strconv"
	"time"

	"github.com/tiket-il/tiket-connect-agent/internal/compare"
	"github.com/tiket-il/tiket-connect-agent/internal/lookup"
	"github.com/tiket-il/tiket-connect-agent/internal/protocol"
)

type vectors struct {
	MatchThreshold int `json:"match_threshold"`
	Signature      []struct {
		Name              string `json:"name"`
		KeyID             string `json:"key_id"`
		Secret            string `json:"secret"`
		Timestamp         string `json:"timestamp"`
		Body              string `json:"body"`
		ExpectedSignature string `json:"expected_signature"`
	} `json:"signature"`
	Normalization []struct {
		Fn       string `json:"fn"`
		Input    string `json:"input"`
		Expected string `json:"expected"`
	} `json:"normalization"`
	Comparison []struct {
		Name     string         `json:"name"`
		Claim    map[string]any `json:"claim"`
		Record   map[string]any `json:"record"`
		Expected struct {
			Confidence      int      `json:"confidence"`
			MatchedFields   []string `json:"matched_fields"`
			UnmatchedFields []string `json:"unmatched_fields"`
			CriticalFailed  bool     `json:"critical_failed"`
			Result          string   `json:"result"`
		} `json:"expected"`
	} `json:"comparison"`
}

func flexString(v any) string {
	switch t := v.(type) {
	case nil:
		return ""
	case string:
		return t
	case float64:
		if t == float64(int64(t)) {
			return strconv.FormatInt(int64(t), 10)
		}
		return strconv.FormatFloat(t, 'f', -1, 64)
	case bool:
		return strconv.FormatBool(t)
	default:
		return fmt.Sprintf("%v", t)
	}
}

func mapToTicket(m map[string]any) compare.Ticket {
	standing, _ := m["is_standing"].(bool)
	return compare.Ticket{
		Barcode:    flexString(m["barcode"]),
		EventName:  flexString(m["event_name"]),
		Artist:     flexString(m["artist"]),
		Venue:      flexString(m["venue"]),
		Date:       flexString(m["date"]),
		Time:       flexString(m["time"]),
		Section:    flexString(m["section"]),
		Row:        flexString(m["row"]),
		Seat:       flexString(m["seat"]),
		IsStanding: standing,
	}
}

// demoLookup is the in-process fake partner for the live-loop checks.
type demoLookup struct {
	tickets       map[string]*lookup.Record
	transferCalls int
	noTransfer    bool
}

func (d *demoLookup) Mode() string            { return "demo" }
func (d *demoLookup) TransferSupported() bool { return !d.noTransfer }
func (d *demoLookup) Lookup(_ context.Context, barcode string) (*lookup.Record, error) {
	return d.tickets[barcode], nil
}
func (d *demoLookup) Transfer(_ context.Context, tr lookup.TransferRequest, _ *lookup.Record) (*lookup.TransferOutcome, error) {
	d.transferCalls++
	return &lookup.TransferOutcome{
		OK:            true,
		NewBarcode:    "REISSUED-" + tr.TransferRef,
		BarcodeFormat: "qr",
		Delivery:      "new_barcode",
	}, nil
}

// Run executes the suite; returns the number of failures.
func Run(vectorsPath string) int {
	failures := 0
	passed := 0
	check := func(name string, ok bool, detail string) {
		if ok {
			passed++
			fmt.Printf("  ✓ %s\n", name)
		} else {
			failures++
			fmt.Printf("  ✗ %s\n    %s\n", name, detail)
		}
	}

	fmt.Printf("tiket-agent self-test (protocol v%s)\n\n", protocol.APIVersion)

	// ── Shared spec vectors ──
	data, err := os.ReadFile(vectorsPath)
	if err != nil {
		fmt.Printf("  ✗ cannot read vectors file %s: %v\n", vectorsPath, err)
		return 1
	}
	var v vectors
	if err := json.Unmarshal(data, &v); err != nil {
		fmt.Printf("  ✗ vectors file is invalid JSON: %v\n", err)
		return 1
	}

	sigOK, sigDetail := true, ""
	for _, s := range v.Signature {
		got := protocol.ComputeSignature(s.Secret, s.Timestamp, []byte(s.Body))
		if got != s.ExpectedSignature {
			sigOK = false
			sigDetail = fmt.Sprintf("%s/%s: got %s want %s", s.KeyID, s.Name, got, s.ExpectedSignature)
			break
		}
	}
	check("spec signature vectors", sigOK, sigDetail)

	normOK, normDetail := true, ""
	impls := map[string]func(string) string{
		"date": compare.NormDate,
		"time": compare.NormTime,
		"text": compare.NormText,
		"seat": compare.NormSeatPart,
	}
	for _, n := range v.Normalization {
		if got := impls[n.Fn](n.Input); got != n.Expected {
			normOK = false
			normDetail = fmt.Sprintf("%s(%q): got %q want %q", n.Fn, n.Input, got, n.Expected)
			break
		}
	}
	check("spec normalization vectors", normOK, normDetail)

	cmpOK, cmpDetail := true, ""
	for _, c := range v.Comparison {
		r := compare.Compare(mapToTicket(c.Claim), mapToTicket(c.Record))
		result := "mismatch"
		if r.Confidence >= v.MatchThreshold && !r.CriticalFailed {
			result = "match"
		}
		if r.Confidence != c.Expected.Confidence ||
			!reflect.DeepEqual(r.Matched, c.Expected.MatchedFields) ||
			!reflect.DeepEqual(r.Unmatched, c.Expected.UnmatchedFields) ||
			r.CriticalFailed != c.Expected.CriticalFailed ||
			result != c.Expected.Result {
			cmpOK = false
			cmpDetail = fmt.Sprintf("%q: got conf=%d matched=%v unmatched=%v crit=%v result=%s; want conf=%d matched=%v unmatched=%v crit=%v result=%s",
				c.Name, r.Confidence, r.Matched, r.Unmatched, r.CriticalFailed, result,
				c.Expected.Confidence, c.Expected.MatchedFields, c.Expected.UnmatchedFields, c.Expected.CriticalFailed, c.Expected.Result)
			break
		}
	}
	check(fmt.Sprintf("spec comparison vectors (%d cases)", len(v.Comparison)), cmpOK, cmpDetail)

	// ── Live handler loop (authenticated envelopes, end to end) ──
	price := 350.0
	demo := &demoLookup{tickets: map[string]*lookup.Record{
		"1000000000001": {
			Barcode: "1000000000001", EventName: "עומר אדם — סיבוב קיץ", Venue: "היכל מנורה מבטחים",
			Date: "2030-08-15", Time: "21:00", Section: "A", Row: "12", Seat: "7",
			Status: "active", OriginalPrice: &price, TicketRef: "DEMO-1",
		},
		"1000000000003": {
			Barcode: "1000000000003", EventName: "הופעה שבוטלה", Venue: "היכל התרבות",
			Date: "2030-10-10", IsStanding: true, Status: "cancelled",
		},
	}}
	handler := &protocol.Handler{Lookup: demo}
	secrets := map[string]string{"k_test": "selftest-secret-0123456789abcdef"}
	ctx := context.Background()

	signedBody := func(payload any) ([]byte, string, string) {
		body, _ := json.Marshal(payload)
		ts := strconv.FormatInt(time.Now().Unix(), 10)
		return body, ts, protocol.ComputeSignature(secrets["k_test"], ts, body)
	}
	authOK := func(body []byte, ts, sig, keyID string) error {
		return protocol.Authenticate(secrets, keyID, ts, sig, body, 300)
	}

	verifyClaim := map[string]any{
		"barcode": "1000000000001", "event_name": "עומר אדם — סיבוב קיץ", "artist": "עומר אדם",
		"venue": "היכל מנורה", "date": "2030-08-15", "time": "21:00",
		"section": "A", "row": "012", "seat": "07", "is_standing": false,
	}

	{
		body, ts, sig := signedBody(map[string]any{"api_version": "1", "request_id": "st-1", "ticket": verifyClaim})
		err := authOK(body, ts, sig, "k_test")
		reply := handler.HandleVerify(ctx, body)
		var out map[string]any
		json.Unmarshal(reply.Body, &out)
		check("signed verify → match (fuzzy seats + venue)",
			err == nil && reply.Status == 200 && out["result"] == "match" && out["original_price"] == 350.0,
			fmt.Sprintf("auth=%v status=%d body=%s", err, reply.Status, reply.Body))
	}

	{
		body, ts, _ := signedBody(map[string]any{"api_version": "1", "ticket": verifyClaim})
		err := authOK(body, ts, "deadbeef", "k_test")
		check("bad signature rejected", err != nil, "expected auth error")
	}

	{
		body, _, _ := signedBody(map[string]any{"api_version": "1", "ticket": verifyClaim})
		staleTs := strconv.FormatInt(time.Now().Unix()-3600, 10)
		sig := protocol.ComputeSignature(secrets["k_test"], staleTs, body)
		err := authOK(body, staleTs, sig, "k_test")
		check("stale timestamp rejected", err != nil, "expected replay-window error")
	}

	{
		body, ts, sig := signedBody(map[string]any{"api_version": "1", "ticket": verifyClaim})
		err := authOK(body, ts, sig, "") // no key-id header → try all secrets
		check("missing key-id falls back to all secrets", err == nil, fmt.Sprintf("%v", err))
	}

	{
		unknown := map[string]any{}
		for k, v := range verifyClaim {
			unknown[k] = v
		}
		unknown["barcode"] = "9999999999999"
		body, _, _ := signedBody(map[string]any{"api_version": "1", "ticket": unknown})
		reply := handler.HandleVerify(ctx, body)
		var out map[string]any
		json.Unmarshal(reply.Body, &out)
		check("unknown barcode → not_found", out["result"] == "not_found", string(reply.Body))
	}

	transferPayload := func(ref, barcode string) []byte {
		body, _ := json.Marshal(map[string]any{
			"api_version": "1", "request_id": "st-t",
			"transfer": map[string]any{
				"barcode": barcode, "transfer_ref": ref,
				"new_holder": map[string]string{"first_name": "דנה", "last_name": "כהן", "email": "buyer@example.com", "phone": "0501234567"},
			},
		})
		return body
	}

	{
		reply := handler.HandleTransfer(ctx, transferPayload("TIKET-TX-1", "1000000000001"))
		var out map[string]any
		json.Unmarshal(reply.Body, &out)
		check("transfer → transferred with barcode_format",
			out["result"] == "transferred" && out["new_barcode"] == "REISSUED-TIKET-TX-1" && out["barcode_format"] == "qr",
			string(reply.Body))
	}

	{
		reply := handler.HandleTransfer(ctx, transferPayload("TIKET-TX-1", "1000000000001"))
		var out map[string]any
		json.Unmarshal(reply.Body, &out)
		check("replayed transfer_ref → idempotent (1 call)",
			out["result"] == "transferred" && demo.transferCalls == 1,
			fmt.Sprintf("calls=%d body=%s", demo.transferCalls, reply.Body))
	}

	{
		reply := handler.HandleTransfer(ctx, transferPayload("TIKET-TX-2", "1000000000003"))
		var out map[string]any
		json.Unmarshal(reply.Body, &out)
		check("transfer of cancelled ticket → rejected",
			out["result"] == "rejected", string(reply.Body))
	}

	{
		bare := &protocol.Handler{Lookup: &demoLookup{tickets: demo.tickets, noTransfer: true}}
		reply := bare.HandleTransfer(ctx, transferPayload("TIKET-TX-3", "1000000000001"))
		var out map[string]any
		json.Unmarshal(reply.Body, &out)
		check("no transfer hook → not_supported", out["result"] == "not_supported", string(reply.Body))
	}

	total := passed + failures
	suffix := ""
	if failures == 0 {
		suffix = " — the agent is ready 🎉"
	}
	fmt.Printf("\n%d/%d checks passed%s\n", passed, total, suffix)
	return failures
}
